Refactoring: define macro rules to declare structs in one place

darknight opened the pull request:
This PR is for issue: https://github.com/slint-ui/slint/issues/3056

It's still in progress, but I want to raise a draft PR to collect feedback before I continue to move forward.

The main changes currently are four macro definitions:
- `for_each_builtin_structs` & `declare_builtin_structs` are the counterpart of enums' `for_each_enums` & `declare_enums`, it is straightforward as it should be
- The most difficult parts are `for_each_export_structs` & `declare_export_structs`
  1) I have to rename fields like `min-width` to `min_width` in `for_each_export_structs` as `min-width` can not be parsed correctly as `ident` in marco arm.
  2) I tried a few ways to parse comments like `//slint::private_api::StateInfo` to get internal struct name but failed, so I have to assume they follow the same prefix, i.e, `slint::private_api::` when I create `Type::Struct` instance in macro `declare_export_structs`
  3) I had thought to merge two `for_each_xxx_structs` macros, but first some struct, for example `KeyEvent` contains unexported fields compared to the one defined in `builtins.slint` file. Second, the type definitions of fields are so different, for example, `TableColumn` which make it very difficult to merge together.
  4) The most tricky part is the instance dependency inside macro `declare_export_structs`. To get rid of loading structs from `builtin.slint` file (by `load_builtins` method), our macro must do this job, so it has to create `Type::Struct` instance for us. But be aware that struct `PointerEvent` & `KeyEvent` contain field `modifiers` which is an instance of struct `KeyboardModifiers`. That means we have to create `KeyboardModifiers` instance first, then refer it during the initialization of other structs. So the marco arm of `declare_export_structs` has a `head` match ($Name0 definition), specifically for matching `KeyboardModifiers` first. This way I succeed in creating the same instance as `load_builtins` method does, something like:
> Struct { fields: {"button": Enumeration(Enumeration { name: "PointerEventButton", values: ["other", "left", "right", "middle"], default_value: 0, node: None }), "kind": Enumeration(Enumeration { name: "PointerEventKind", values: ["cancel", "down", "up"], default_value: 0, node: None }), "modifiers": Struct { fields: {"alt": Bool, "control": Bool, "meta": Bool, "shift": Bool}, name: Some("slint::private_api::KeyboardModifiers"), node: None, rust_attributes: None }}, name: Some("slint::private_api::PointerEvent"), node: None, rust_attributes: None }

But above handling leads me a few questions & doubts:
1. **Am I on the correct path to complete this task?**
2. Is there any more elegant way to implement `declare_export_structs` macro?
3. Compare to `load_builtins` method, is macro a better way?

I'm still new to slint, any comment is welcome.

ogoffart commented:
Thanks for the PR!
I think you are going very much in the right direction.

Although I had in mind to merge the for_each_export_structs and declare_builtin_structs in one single macro.  Probably keep only for_each_export_structs and get rid of declare_builtin_structs.  Because the for_each_export_structs already has all the information needed. The only difference being the type,  but doing a match like you did with the different type is a valid way.

Responding to your comments:


>    I have to rename fields like min-width to min_width in for_each_export_structs as min-width can not be parsed correctly as ident in marco arm.

Yes that's fine. We can just do replace `_`->`-` when adding them to the type register.

>    I tried a few ways to parse comments like //slint::private_api::StateInfo to get internal struct name but failed, so I have to assume they follow the same prefix, i.e, slint::private_api:: when I create Type::Struct instance in macro declare_export_structs

You can use some special matching in the macro like `@name ="slint::private_api::StateInfo"`  or `#[name(slint::private_api::StateInfo)]`  or whatever.  We are not limited to pure rust syntax although it's nice to keep close to it.

>    I had thought to merge two for_each_xxx_structs macros, but first some struct, for example KeyEvent contains unexported fields compared to the one defined in builtins.slint file. Second, the type definitions of fields are so different, for example, TableColumn which make it very difficult to merge together.

maybe add something like `#[hidden]` or `pub` to mean exported in slint or a `private` keyword to mean not exported in slint.

>    The most tricky part is the instance dependency inside macro declare_export_structs. To get rid of loading structs from builtin.slint file (by load_builtins method), our macro must do this job, so it has to create Type::Struct instance for us. But be aware that struct PointerEvent & KeyEvent contain field modifiers which is an instance of struct KeyboardModifiers. That means we have to create KeyboardModifiers instance first, then refer it during the initialization of other structs. 

Isn't it enough to but the macro in the right order and add them one by one in the type register? What's what we used to do. 

> Compare to load_builtins method, is macro a better way?

That's a good question.  The whole point is to have the information in only one place, but if it is in defferent macro, then it's not really worth it.  I hope one can manage to make it in one single macro.

darknight commented:
Thanks for the reply @ogoffart, those suggestions are very helpful.

By slightly changing the format of structs listed in `for_each_builtin_structs` macro, I managed to have **one** single definition (although it doesn't fully respect rust syntax)

As for this point:
> > The most tricky part is the instance dependency inside macro declare_export_structs. To get rid of loading structs from builtin.slint file (by load_builtins method), our macro must do this job, so it has to create Type::Struct instance for us. But be aware that struct PointerEvent & KeyEvent contain field modifiers which is an instance of struct KeyboardModifiers. That means we have to create KeyboardModifiers instance first, then refer it during the initialization of other structs.
> 
> Isn't it enough to but the macro in the right order and add them one by one in the type register? What's what we used to do.

It may not be enough for this case. The problem is because of the type conversion during the instance initialization.
If apply the **universal** match as we did for other cases, the code may look like:
```
                $(let $Name = Type::Struct {
                    fields: BTreeMap::from([
                        $((stringify!($pub_field).replace('_', "-"), match stringify!($pub_type) {
                            "bool" => Type::Bool,
                            "i32" => Type::Int32,
                            "f32" => Type::Float32,
                            "SharedString" => Type::String,
                            "Coord" => Type::LogicalLength,
                            "KeyboardModifiers" => KeyboardModifiers,  // *<- this is the problem
                            _ => BUILTIN_ENUMS.with(|e| e.get_enum(stringify!($pub_type))),
                        })),*
                    ]),
                };)*
```
Since `KeyboardModifiers` is used by other struct, it appears in the type conversion, which means it appears during the initialization of  `KeyboardModifiers` also. So the code generated could be:
```
        let KeyboardModifiers = Type::Struct {
            fields: BTreeMap::from([
                (stringify!( alt ).replace('_', "-"), match stringify!( bool ) {
                    "bool" => Type::Bool,
                    "i32" => Type::Int32,
                    "f32" => Type::Float32,
                    "SharedString" => Type::String,
                    "Coord" => Type::LogicalLength,
                    "KeyboardModifiers" => KeyboardModifiers,  // *<- this is the problem
                    _ => BUILTIN_ENUMS.with(|e| e.get_enum(stringify!( bool ))),
                }),
                // omitted
            ]),
            // omitted
        };
```
We refer `KeyboardModifiers` during the initialization of `KeyboardModifiers`.

That's why I split the macro arm into 2 parts: one match for `KeyboardModifiers` specifically, one match for the rest of structs. And `match stringify!($pub_type)` statement is slightly different for each part during the initialization.

I understand this might not be ideal, so any ideas are huge welcome.

ogoffart commented on internal/compiler/typeregister.rs:
```diff
@@ -65,14 +65,110 @@ macro_rules! declare_enums {
+macro_rules! declare_export_structs {
+    (
+        // match `KeyboardModifiers` first since it's dependent by following struct instances
+        $(#[$attr0:meta])*
```
Why do you need to duplicate this in a 0 and rest version?

ogoffart commented on internal/compiler/typeregister.rs:
```diff
@@ -65,14 +65,110 @@ macro_rules! declare_enums {
+                            "f32" => Type::Float32,
+                            "SharedString" => Type::String,
+                            "Coord" => Type::LogicalLength,
+                            _ => BUILTIN_ENUMS.with(|e| e.get_enum(stringify!($pub_type0))),
```
did `e.$pub_type0` not work?

ogoffart commented on internal/compiler/typeregister.rs:
```diff
@@ -65,14 +65,110 @@ macro_rules! declare_enums {
+                let $Name0 = Type::Struct {
+                    fields: BTreeMap::from([
+                        $((stringify!($pub_field0).replace('_', "-"), match stringify!($pub_type0) {
+                            "bool" => Type::Bool,
+                            "i32" => Type::Int32,
+                            "f32" => Type::Float32,
+                            "SharedString" => Type::String,
+                            "Coord" => Type::LogicalLength,
```
Maybe here, instead of runtime matching, it can be compile time with a macro.

something like
```
macro_rules! map_type {
  (bool) => { Type::Bool };
  (i32) => { Type::Int32 };
  ...
}
```

darknight replied:
Thanks for the hint!

After a bit of refactoring, finally I eliminated special match for `KeyboardModifiers`, make the macro `declare_export_structs` general declaration for builtin structs.

Please have a look one more time, appreciate!

darknight commented on internal/compiler/typeregister.rs:
```diff
@@ -71,8 +71,77 @@ macro_rules! declare_enums {
+        Type::LogicalLength
+    };
+    ($pub_type:ident, KeyboardModifiers) => {
+        $pub_type.clone()
```
btw, writting `KeyboardModifiers.clone()` directly is not working, it'll cause compile error:
> a field by this name exists in `Self`

That's why I pass in the original `$pub_type:ident` as first parameter.

ogoffart reviewed (commented):
Great.

The next step is to generate the C++ header  in `api/cpp/cbindgen.rs` in the same way as the enums

darknight commented:
> Great.
> 
> The next step is to generate the C++ header in `api/cpp/cbindgen.rs` in the same way as the enums

@ogoffart Thanks for the direction. I managed to generate new cpp header for those builtin structs but faced a few issues so need your help.

- `StandardListViewItem` is in the namespace `slint` not `slint::cbindgen_private` on the **master** branch, my change removed it from `slint_generated_public.h` file, and moved into `slint::cbindgen_private` namespace in `slint_builtin_structs_internal.h` file. I only saw `api/cpp/include/slint_image.h` directly includes `slint_generated_public.h`, but not using `StandardListViewItem`, so I suppose this change should be compatible.
- `StateInfo` on **master** branch defines type alias, like:
  > struct StateInfo {
  > using Instant = uint64_t;
  > Instant change_time;
  > }

  my change use `uint64_t` directly since now `StateInfo` is declared from macro, and this type alias is too specific to one struct. Again I suppose it's a compatible change.
- `StandardListViewItem` & `TableColumn` overloads **==** operator as friend function, something like:
  > friend bool operator==(const TableColumn&, const TableColumn&) = default;

  `StateInfo` overloads **==**, **!=** as member functions, like:
  > bool operator==(const StateInfo& other) const {  return current_state == other.current_state &&  previous_state == other.previous_state &&  change_time == other.change_time;  }  
bool operator!=(const StateInfo& other) const {  return current_state != other.current_state ||  previous_state != other.previous_state ||  change_time != other.change_time;  }

  They're all on **master** branch.
  This is the part I've no idea how to handle it in the macro. They're specific to some struct with its own logic, but the macro expects to generate common logic, so need your advice to take care of this issue.

ogoffart commented:
> StandardListViewItem is in the namespace slint not slint::cbindgen_private on the master branch, my change removed it from slint_generated_public.h file, and moved into slint::cbindgen_private namespace in slint_builtin_structs_internal.h file. I only saw api/cpp/include/slint_image.h directly includes slint_generated_public.h, but not using StandardListViewItem, so I suppose this change should be compatible.

I recently wanted to also add an enum as part of the public API, so i made a change to generate both a public and a non-public header for enums. See commit 3a2359f107b1d2e008089465326246952aedab26
You can do the same for structs.

> my change use uint64_t directly since now StateInfo is declared from macro, and this type alias is too specific to one struct. Again I suppose it's a compatible change.

Yes that's fine.

> operator==

You can use the `= default`,  that's the C++20 way to do it.  and it should be compatible with the old way.

ogoffart commented on internal/core/model.rs:
```diff
@@ -8,6 +8,7 @@
 use crate::component::ComponentVTable;
 use crate::item_tree::TraversalOrder;
 use crate::items::{ItemRef, SortOrder};
```
SortOrder is no longer used and cause a warning which blocks CI

ogoffart commented on api/cpp/cbindgen.rs:
```diff
@@ -49,6 +49,57 @@ fn enums(path: &Path) -> anyhow::Result<()> {
+        .context("Error creating slint_builtin_structs_internal.h file")?;
+    writeln!(structs, "#pragma once")?;
+    writeln!(structs, "// This file is auto-generated from {}", file!())?;
+    writeln!(structs, "namespace slint::cbindgen_private {{")?;
```
Before this line, you also need to include the header for generated enums, i think

ogoffart commented on internal/common/builtin_structs.rs:
```diff
@@ -0,0 +1,115 @@
+
+//! This module contains all builtin structures.
+
+/// TODO: add doc and example
```
yes, a doc would be nice.

ogoffart commented on internal/compiler/typeregister.rs:
```diff
@@ -71,8 +71,77 @@ macro_rules! declare_enums {
+
 thread_local! {
     pub static BUILTIN_ENUMS: BuiltinEnums = BuiltinEnums::new();
+    pub static EXPORT_STRUCTS: ExportStructs = ExportStructs::new();
```
```suggestion
    static EXPORT_STRUCTS: ExportStructs = ExportStructs::new();
```

I don't think anyone uses that currently so let's not make it pub unless it is needed

ogoffart reviewed (commented):
Thanks.

Can you please rebase your branch and squash commits?

I thin kit is nice improvements, but there is still things that would be nice to use this macro for so we declare all the enums only in one place:

 - documentation, The goal is to generate automatically  docs/language/src/builtins/structs.md  see: https://github.com/slint-ui/slint/blob/e829564e8a866be7a428751d4768a684f187491d/xtask/src/slintdocs.rs#L109

 - change declare_value_struct_conversion so that we can use that macro instead of manually keeping it in sync: https://github.com/slint-ui/slint/blob/e829564e8a866be7a428751d4768a684f187491d/internal/interpreter/api.rs#L255-L262

 
 - internal/core/rtti.rs  so we don't have to repeat each struct name in declare_ValueType_2. Since now we need inputs from two macros, we might have to turn macro around.

 - bonus: We could use the actual type instead of Value for kown struct in https://github.com/slint-ui/slint/blob/e829564e8a866be7a428751d4768a684f187491d/internal/interpreter/dynamic_component.rs#L983-L986
 
 
 The documentation and declare_value_struct_conversion are the most important because they need to be kept in sync manally right now for every field that changes, and we don't get test faillure for the docs so we tend to forget it. and we don't get test faillure for declare_value_struct_conversion either until we really use that field in the tests for the interpreter.

darknight commented:
@ogoffart Thanks for the suggestions.

Currently, I managed to implement **doc generation** & **value struct conversion** by applying the new macro I created.

For the improvement of `declare_ValueType_2` and `dynamic_component.rs`, if it's not so important, I suggest to create separate issue to track them.

One more thing I need to point out, the exiting `structs.md` doc contains `Point` struct, which is not listed as builtin struct in current implementation. After applying the logic here, the `Point` will be removed, and `StateInfo` will be added into generated `structs.md` file.

ogoffart commented:
> After applying the logic here, the Point will be removed, and StateInfo will be added into generated structs.md file.

StateInfo should not be there in the documentation. So you should find a way to solve this, either by removing it from the TreeMap manually before printing, or some other trick.
Same for the `Point`, we must keep it in the doucmentation.
Adding it manually in the TreeMap if there isn't another way.

darknight commented:
> > After applying the logic here, the Point will be removed, and StateInfo will be added into generated structs.md file.
> 
> StateInfo should not be there in the documentation. So you should find a way to solve this, either by removing it from the TreeMap manually before printing, or some other trick. Same for the `Point`, we must keep it in the doucmentation. Adding it manually in the TreeMap if there isn't another way.

Thansk for the review, I resolved this comment. Now the change here will generate almost identical `structs` documentation as before.

Btw, for the improvement of `declare_ValueType_2`, I played a bit and found it's quite tricky, currently still not figure out an elegant solution, so I might not include the other two improvements you mentioned before. Perhaps in a separate PR in future.

darknight commented:
@ogoffart Hi, any more comments for this change?

ogoffart commented:
Your work on this refactor has been impressive, and I'm pleased with how well it has shaped up. I appreciate your dedication in addressing all the comments.
The final version of the PR looks great! Your changes have improved the codebase. So i'm merging it.

Thanks again!

darknight commented:
> Your work on this refactor has been impressive, and I'm pleased with how well it has shaped up. I appreciate your dedication in addressing all the comments. The final version of the PR looks great! Your changes have improved the codebase. So i'm merging it.
> 
> Thanks again!

Hi @ogoffart, thanks for the merge! I l'm glad my work is helpful on slint codebase. I like it & I'll continue to contribute :)
